src/HttpServer.php

```diff
@@ -60,50 +64,115 @@ public function addMiddleware($middleware)
 
-        $chunkStream = new ReadableStream();
-        $chunkedDecoder = new ChunkedDecoder($chunkStream);
 
-        $headerStream = new ReadableStream();
-        $headerDecoder = new HeaderDecoder($headerStream);
+        $connection->on('data', function ($data) use (&$connection, $that, &$request, &$headerBuffer) {
+            $headerBuffer .= $data;
+            if (strpos($headerBuffer, "\r\n\r\n") !== false) {
+                // header is completed
+                $fullHeader = substr($headerBuffer, 0, strpos($headerBuffer, "\r\n\r\n") + 4);
 
-        $connection->on('data', function ($data) use ($connection, &$headerCompleted, &$bodyBuffer, $that, &$chunkedDecoder, &$headerDecoder, $chunkStream, $headerStream) {
-            if (!$headerCompleted) {
-                $headerDecoder->on('data', function ($header) use (&$request, &$data, &$headerCompleted) {
-                    $request = RingCentral\Psr7\parse_request($header);
-                    $data = substr($data, strlen($header));
-                    $headerCompleted = true;
-                });
-                $headerStream->emit('data', array($data));
+                try {
+                    $request = RingCentral\Psr7\parse_request($fullHeader);
+                } catch (\Exception $ex) {
+                    $that->sendResponse(new Response(400), $connection);
+                    return;
+                }
+                // remove header from $data, only body is left
+                $data = (string)substr($data, strlen($fullHeader));
+
+                $that->handleBody($request, $connection);
+                $connection->emit('data', array($data));
             }
+        });
+    }
 
-            if (isset($request)) {
-                if ($that->isChunkedEncodingActive($request)) {
-                    $chunkedDecoder->on('data', function ($chunk) use (&$bodyBuffer, &$request, $that, $connection) {
-                        $bodyBuffer .= $chunk;
-                        if (strlen($chunk) == 0) {
-                            $that->sendBody($bodyBuffer, $connection, $request);
-                        }
-                    });
-                    $chunkStream->emit('data', array($data));
-                } else {
-                    $bodyBuffer .= $data;
-                    $contentLengthArray = $request->getHeader('Content-Length');
-
-                    if (!empty($contentLengthArray) && strlen($bodyBuffer) == $contentLengthArray[0]) {
-                        $that->sendBody($bodyBuffer, $connection, $request);
-                    } else if (empty($contentLengthArray) || $contentLengthArray[0] == 0) {
-                        $that->handleRequest($connection, $request);
-                    }
+    /** @internal */
+    public function handleBody(RequestInterface $request, ConnectionInterface $connection)
+    {
+        $input = $this->sendHttpBodyStream($request, $connection);
+
+        $connection->removeAllListeners('data');
```

Comment:
What listener is being removed here? Who added this listener in the first place? What if there are other listeners?